### ISSUE TITLE
Incubation of JMH engine

### ISSUE THREAD

**Issue by mp911de** [junit-pioneer/junit-pioneer#138]
We created a JUnit 4 runner/JUnit 5 `TestEngine` while working on JMH benchmarks and found it quite useful during development-time of JMH benchmarks. What is basically does, is configuring JMH's runner scoped to a benchmark class/method and report back the progress.

We found this is not only useful for ourselves but there's also interest from the community. However, it's not a typical Spring project so we’re trying to find a better home for it.

The code is EPL licensed and we’re interested in driving development. What do you think about hosting the code at the JUnit Pioneer org in a separate repository?

I think the code is only useful as long there is no proper IDE support available and JUnit is a well-established technology to bridge execution requests from the IDE to JMH's runner.

You can find the code at: https://github.com/mp911de/microbenchmark-runner

/cc @christophstrobl @dsyer @odrotbohm

**Comment by dsyer** [comment]
I'd be happy to continue contributing here (as opposed to Mark's personal space). I'd also be fine with only supporting JUnit 5, if that makes any difference.

**Comment by dsyer** [comment]
> I think the code is only useful as long there is no proper IDE support available

I'm not sure I agree with that. I would use the microbenchmark runner for JUnit 5 even if JMH had better IDE support, unless it worked exactly the same way (i.e. only used test-scoped dependencies, and could be hooked into the integration test phase of the build). I'm not sure it would be worth anyone's while to provide such support when JUnit 5 has already such a nice extension API.

**Comment by smoyer64** [comment]
I'll start by saying that I think this is a very cool project but I have some concerns.  I understand why the code is structured as a ``TestEngine`` ... the common code is (was?) based on a ``Runner``.  I've changed the title to reflect this.  In general, the problem with ``Runner`` is that you could only assign one per test class (or suite).  There are many more extension points in JUnit 5 and I'd be curious to see if this could in fact be an ``Extension``.  With the ``@MicroBenchmark`` and ``@Benchmark`` annotations - and looking at the example benchmark in the README - it's clear that you don't intend for there to be an intersection of "executions" that are both ``@Test`` and ``@Benchmark`` so maybe this is a non-problem.

As a ``TestEngine``, I don't think this project should be merged into the JUnit Pioneer extension library.  I wouldn't be opposed to this becoming a project within the JUnit Pioneer group but I should warn you that we'd have to do some work to publish this to Maven Central and to organize the documentation so that there are clear sections for separate projects.  In the longer term, I think it's inevitable that the JUnit 5 community will want to code that doesn't fit into the core library so this is only a problem because we haven't entirely worked out how we want to manage these projects.

I wouldn't be opposed to accepting the project here and adding the existing committers to the project's members list - and I see that @sbrannen has already committed.

Opinions @junit-pioneer-team?

**Comment by mp911de** [comment]
I get the feeling that dropping JUnit 4 support (`Runner`) seems a reasonable move. I also think this should be a separate project.

I had a chat with @sbrannen and his suggestion was a custom `TestEngine`. This way, we prevent lingering instances that are disconnected from the actual JMH execution.

**Comment by mp911de** [comment]
@nicolaiparlog Did you get a chance to look into this one?

**Comment by smoyer64** [comment]
@mp911de @dsyer 

As I mentioned above, it's my opinion that this should remain a separate project but I'd be happy with it being a JUnit Pioneer project.  If you want to transfer the project to the junit-pioneer organization, I'll add you both as maintainers.  We'll need to discuss what documentation is required and how we'll merge the documentation for multiple projects into a single web-site.

@nicolaiparlog 

Can you comment on how you envision the web-site changing for multiple projects? (I know we discussed this 18 months ago or so but it's not in my notes).

**Comment by smoyer64** [comment]
Actually I see that @nicolaiparlog has documented how the site is built in the README at https://github.com/junit-pioneer/junit-pioneer.github.io.  I'm not sure I understand the process completely but it appears that the first step would be to create asciidoc pages in the /doc folder of your project.  (like https://github.com/junit-pioneer/junit-pioneer/tree/master/docs).

**Comment by mp911de** [comment]
> If you want to transfer the project to the junit-pioneer organization, I'll add you both as maintainers. 

Sounds good.

> We'll need to discuss what documentation is required and how we'll merge the documentation for multiple projects into a single web-site.

Let me know what's required from my side and which groupId we should use for the Microbenchmarkl runner. Right now, it's `com.github.mp911de.microbenchmark-runner:microbenchmark-runner-core`/`com.github.mp911de.microbenchmark-runner:microbenchmark-runner-junit4`/`com.github.mp911de.microbenchmark-runner:microbenchmark-runner-junit5`.

**Comment by smoyer64** [comment]
I kind of like the way you phrased it in the title.  I think we should "reserve" ``org.junit-pioneer`` for the extensions project so how does ``org.junit-pioneer.jmh-engine`` sound for a groupId?

**Comment by nicolaiparlog** [comment]
Okay, just 10 months after Mark's initial request, I'm finally to here to take a look. Sorry to have kept you waiting. :blush: My opinions in short:

* the project is really interesting and I would love to make it part of JUnit Pioneer :smiley: 
* I agree that this specific extension is better well-suited to become its own project under Pioneer's roof than part of the Pioneer Expansion Pack
* I agree that this should be an engine, because it not only doesn't need interaction with Jupiter features, it should actively prevent them, so they can't creep in accidentally

@mp911de @dsyer If you're still interested, I'll make you maintainers, so you can transfer the project. I can then help you let the documentation show up in the right place.

**Comment by mp911de** [comment]
No worries, we weren't just sitting around and waiting. I also felt kind of guilty of not pushing things.

How about the following steps:

1. You create a repo in this org
2. We're going to add all that code as PR so we all get a chance to review and align the code before it gets into `master`
3. I'm fine being a maintainer for the new repo.

This way, we have a clear IP transfer and once this is done, I'm happy to archive the original repo at https://github.com/mp911de/microbenchmark-runner.

**Comment by nicolaiparlog** [comment]
There you go (quick as ever :grin:): https://github.com/junit-pioneer/microbenchmark-runner

I've invited you two to become maintainers, but I like the idea with vetting the code via PR. Once we've done that, I will create a PR to pull in the docs.

**Comment by nicolaiparlog** [comment]
All work is done here, right? If not, let me know and I'll reopen.
